### PR TITLE
Add filtering functionality to the symbols list

### LINKED ISSUE #5761: Punctuation dialog needs a way to search symbols
Can the list of symbols in the punctuation/symbol pronunciation settings have a search feature?
Earlier I was trying to find the Ø symbol so I could change it's setting, but I have a very long list of characters and I'm not sure how to type it on the keyboard. If I could copy and paste it and search the list for matches I could probably find it faster?

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #5761

### Summary of the issue:
Now that emoji's have been introduced in NVDA symbol dictionaries, the list of symbols was getting pretty big. This also introduced a major lag when opening the speech symbols dialog.

### Description of how this pull request fixes the issue:
1. Added a filter edit box to the speech symbols dialog. @feerrenrut: your opinion about the visuals of this edit box are greatly appreciated. Feel free to push directly to this branch if you need to change things visually.
2. Converted the list control to a list control with the wx.LC_VIRTUAL style. I'm really exited about the major performance boost this gives when loading the dialog for the first time, as well as when filtering. It is just really fast now!
3. Added some functionality to gui.nvdaControls.AutoWidthColumnListCtrl to support getting item text from a data storage without the requirement of subclassing the list control. It is now also possible to forcefully send a wx.EVT_LIST_ITEM_FOCUSED to the control

### Testing performed:
Filtering, adding and removing items, closing the dialog and reopening it again quickly after that.

### Known issues with pull request:
* The visuals might not be ideal yet, though guiHelper is really amazing in creating stuf that just looks good.
* We should probably wait for #8006 to be in master and apply scaling support to this dialog as well while at it.
* This does not yet apply to speech dictionaries. I think it should at some point, but that won't be part of this pr.

### Change log entry:
* New features
    + You can now filter symbols in the punctuation/symbol pronunciation dialog, similar to how filtering works in the elements list and input gestures dialog. (#5761)

* Bug fixes
   + Loading the punctuation/symbol pronunciation dialog is now much faster when using symbol dictionaries containing over 1000 entries.

### CODE REVIEW

**Comment by leonardder** [comment]
Wow! I'm coming across like 31000 symbols here, taking like 20 seconds to load with the current Alpha version of NVDA. This pr reduces that to 2 seconds.

**Comment by dnz3d4c** [comment]
Thanks.
> 
> 
> Wow! I'm coming across like 31000 symbols here, taking like 20 seconds to load with the current Alpha version of NVDA. This pr reduces that to 2 seconds.

**Comment by dingpengyu** [comment]
@Hi leonardder I am the maintainer of NVDA Simplified Chinese. I found that NVDA can't generate Simplified Chinese emoticon dictionary using the scons command.

**Comment by josephsl** [comment]
Hi, that’ll be because it isn’t clear at the first glance which XML file is for which Chinese variant. This is something that should be addressed sometime in October. Thanks.

**Comment by dingpengyu** [comment]
thank josephsl

**Comment by leonardder** [comment]
I must have missed the Chinese annotations.

@dingpengyu: Could you please open a new issue for this? I'm not yet known with the various differences in the Chinese language, so it is quite difficult to decided what would be the best assignments. In NVDA, we have the zh-cn, zh-hk and zh-tw locales.

**Comment by josephsl** [comment]
Hi, I can assist with this, as I work with Asian languages in a regular basis. When creating this new issue, please at mention me (CC me). Thanks.

**Comment by dingpengyu** [comment]
josephsl@

**Comment by dingpengyu** [comment]
Hi, thanks to josephsl's reply, this is the issue address:
NVDA cannot generate Simplified Chinese emoticon dictionary using the scons command · Issue #8799
https://github.com/nvaccess/nvda/issues/8799

**Comment by leonardder** [comment]
@feerrenrut: I've removed the blocked label since #8006 is merged. However, I'm not yet sure what to do to make this dialog scale correctly, other than setting resizable to True. The size of the list control and its columns is undefined.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
When filtering, the list does not update. The entries only update as they gain focus. Visually this is quite confusing.

Because this dialog is a `SettingsDialog` Then there is no need to do anything to get the DPI scaling, except, if you specify a size manually, you must pass it through `self.scaleSize()`. 

I think this dialog could be wider, because the text in the replacement column is often long enough to exceed the width of the column. The replacement field (once populated) often isn't wide enough. There is plenty of space to widen this field, and the dialog itself isn't very wide either.

Some emoji don't seem to be supported by the font. Visually they are just squares. Not a blocking issue right now, since I believe this is already the case. An issue should be created for this.

I can take a look at these points.

**Comment by leonardder** [comment]
> When filtering, the list does not update. The entries only update as they gain focus. Visually this is quite confusing.

Ugh, might this be because the list is virtual? Does this also mean that there's only one item visible at a time or something like that?

File: source/gui/settingsDialogs.py
Comment (leonardder) [reply source/gui/settingsDialogs.py]: Mm, this makes sense, but might be a bit out of scope for this pr. See also https://www.python-course.eu/levenshtein_distance.php . it seems to need a new helper function.

File: source/gui/nvdaControls.py
Comment (leonardder) [reply source/gui/nvdaControls.py]: Agreed, though it was not as simple as I thought before.

**Comment by leonardder** [comment]
@feerrenrut: Thanks for fixing the list updating issue. Would be great if you can fix the replacement column width as well, since it would really help to do that with a pair of eyes.

File: source/gui/nvdaControls.py
Comment (feerrenrut) [inline source/gui/nvdaControls.py]: Another option here is to bind to the window destroy event and clean up.

```
@@ -50,21 +49,27 @@ class AutoWidthColumnListCtrl(wx.ListCtrl, listmix.ListCtrlAutoWidthMixin):
 			if not isinstance(itemTextCallable, Callable):
 				raise TypeError("itemTextCallable should be None or a callable")
 			# We need to store a weak reference to the callable.
 			# If we don't and the callable is a bound instance method of a wx Window,
 			# we might create a circular reference.
-			self._itemTextCallable = getWeakrefForCallable(itemTextCallable)
+			self._itemTextCallable = itemTextCallable
 		else:
-			self._itemTextCallable = None
+			self._itemTextCallable = self._super_itemTextCallable
 		wx.ListCtrl.__init__(self, parent, id=id, pos=pos, size=size, style=style)
 		listmix.ListCtrlAutoWidthMixin.__init__(self)
 		self.setResizeColumn(autoSizeColumnIndex)
+		self.Bind(wx.EVT_WINDOW_DESTROY, source=self, id=self.GetId, handler=self._onDestroy)
+
+	def _onDestroy(self, evt):
+		evt.Skip()
+		self._itemTextCallable = None
+

+	def _super_itemTextCallable(self, item, column):
+		return super(AutoWidthColumnListCtrl, self).OnGetItemText(item, column)

 	def OnGetItemText(self, item, column):
-		if self._itemTextCallable is None or self._itemTextCallable() is None:
-			return super(AutoWidthColumnListCtrl, self).OnGetItemText(item, column)
-		return self._itemTextCallable()(item, column)
+		return self._itemTextCallable(item, column)

 	def sendListItemFocusedEvent(self, index):
 		evt = wx.ListEvent(wx.wxEVT_LIST_ITEM_FOCUSED, self.Id)
```

**Comment by feerrenrut** [comment]
Yep, I'll take a look at the widths in the dialog tomorrow.

File: source/gui/nvdaControls.py
Comment (leonardder) [reply source/gui/nvdaControls.py]: This works as well.

File: source/gui/guiHelper.py
Comment (leonardder) [inline source/gui/guiHelper.py]: Is this intentional?

**Review by feerrenrut — APPROVED** [review]
I've looked through this again and I'm happy for it to be merged. @leonardder Are you happy with the latest changes?

File: source/gui/guiHelper.py
Comment (leonardder) [inline source/gui/guiHelper.py]: Could you give this a doc string that clarifies why this subclass has been created?

File: source/gui/guiHelper.py
Comment (leonardder) [inline source/gui/guiHelper.py]: What is this proportion=1? Could a constant help here?

**Review by LeonarddeR — COMMENTED** [review]
I've got some comments. Also, you could consider adding a convenience wrapper for the new LabeledTextControl to BoxSizerHelper.

File: source/gui/guiHelper.py
Comment (feerrenrut) [reply source/gui/guiHelper.py]: Proportion is an argument to wx box sizer, it lets you control how much of the sizers orientation a certain element gets. So for a horizontal sizer, it controls the proportion of width allocated to that particular control. Approximately, it seems that the proportion of any unused space is divided amongst the controls that was added to the sizer with a proportion value, based on the total of all the proportion values. I'm not sure that a constant really makes sense. In this particular case, I think any value has the same effect as any other value, though I'm far from certain about this.

File: source/gui/guiHelper.py
Comment (feerrenrut) [reply source/gui/guiHelper.py]: Yes, I'll do this. Trying to achieve this small variation has made me think carefully about the implementation here. I think we should start to move this towards a more component driven implementation. The initial goal was to try to streamline the GUI code, making it simpler (and less error prone) to do the common things, I think this has been successful. Though now I would like to make this a little more flexible too.

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: The interesting thing here is that column with index 2 is level, not replacement. See below.

Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: Hmm yes, index is really not a good name for this.

**Comment by dingpengyu** [comment]
hi
Is there a developer review for this PR? Seeing that it is already approved

File: source/gui/guiHelper.py
Comment (feerrenrut) [reply source/gui/guiHelper.py]: I decided to remove this class, trying to find a way to implement this variation in a clean way has inspired me to take another approach. For now, this means that two textCtrls that should grow with the dialog will not. I don't think this PR should be blocked on this.

**Comment by feerrenrut** [comment]
@leonardder I've made a few changes here, if you are happy with this I think it is ready to be merged.

**Comment by DataTriny** [comment]
I didn't mean to make any UI design work on this PR. I'm happy with the code, let's wait for Leonard.

**Comment by leonardder** [comment]
@feerrenrut: I'm ok with your changes and did a small round of tests.

Found and fixed the following:

1. Add a new symbol, for example called `test`
2. Filter on test
3. Remove the test symbol

What happened is that the list of symbols was empty, yet the old logic tried to select a new symbol. Removing a symbol didn't check for the item count properly. Now it does.

**Review by feerrenrut — APPROVED** [review]
Looks good. I'll merge this in.

**Pull request merged**
